### PR TITLE
Refactor debounce from utils

### PR DESCRIPTION
**Pull request by vhargrave**
Refactors debounce away from utils as requested here https://github.com/adobecom/milo/pull/711 

Resolves: [MWPW-133591](https://jira.corp.adobe.com/browse/MWPW-133591)

Test URLs:
Test search at top right in new gnav (debounce should be working)
Before: https://main--milo--adobecom.hlx.page/drafts/vhargrave/empty-page?martech=off&hideGeorouting=on
After: https://vhargrave-remove-debounce-utils--milo--adobecom.hlx.page/drafts/vhargrave/empty-page?martech=off&hideGeorouting=on

Test search at top right in old gnav (debounce should be working)
Before: https://main--bacom--adobecom.hlx.page/?martech=off&hideGeorouting=on
After: https://main--bacom--adobecom.hlx.page/?milolibs=vhargrave-remove-debounce-utils&martech=off&hideGeorouting=on

Before: https://main--milo--adobecom.hlx.page/tools/commerce?martech=off&hideGeorouting=on
After: https://vhargrave-remove-debounce-utils--milo--adobecom.hlx.page/tools/commerce?martech=off&hideGeorouting=on

### CODE REVIEW

**Review by mokimo — APPROVED** [review]
What's the idea for the `action` file - should this also grow or just be specific to debounce?

**Comment by vhargrave** [comment]
> What's the idea for the `action` file - should this also grow or just be specific to debounce?

I've named it so that it can grow if there are other util methods for on actions, but we'll see.

File: test/blocks/commerce/commerce.test.js [hunk 1]
```diff
@@ -76,6 +76,27 @@ describe('decorateSearch', () => {
...
+    const e = new KeyboardEvent("keyup", {bubbles : false, cancelable : true, key : "Q", shiftKey : false});
+    Object.defineProperty(e, 'target', {
+      value: {
+        value: 'Q',
+      }
+    });
+    input.dispatchEvent(e);
+    let h4 = el.querySelector('h4');
+    expect(h4).to.equal(null);
+    setTimeout(() => {
```
Comment (overmyheadandbody) [inline test/blocks/commerce/commerce.test.js:95]: I'm not sure if we should use timeouts in tests. Once might be okay, but if we start adding multiple timeouts, the test suite will take longer than needed to run. Maybe something like [this](https://github.com/adobecom/milo/blob/main/test/blocks/global-navigation/global-navigation.test.js#L158-L161) can help.

File: test/utils/action.test.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import { expect } from '@esm-bundle/chai';
+import { debounce } from "../../libs/utils/action.js";
+
+
+describe('Input', () => {
```
Comment (overmyheadandbody) [inline test/utils/action.test.js:5]: Should this be renamed to 'Action'?

File: test/blocks/commerce/commerce.test.js (hunk 1 above)
Comment (vhargrave) [reply test/blocks/commerce/commerce.test.js]: great input, thank you!

File: libs/blocks/commerce/commerce.js [hunk 3]
```diff
@@ -129,18 +130,21 @@ export const decorateOfferDetails = async (el, of, searchParams) => {
 export const handleSearch = async (event, el) => {
   let searchParams = {};
   const displaySearchError = () => {
-    const notValidUrl = document.createElement('h4');
-    notValidUrl.classList.add('not-valid-url');
-    notValidUrl.textContent = 'Not a valid offer link';
+    const notValidUrl = createTag('h4', { class: 'not-valid-url' }, 'Not a valid offer link')
```
Comment (narcis-radu) [inline libs/blocks/commerce/commerce.js:133]: you're missing a `;`

File: libs/blocks/commerce/commerce.js [hunk 4]
```diff
@@ -129,18 +130,21 @@ export const decorateOfferDetails = async (el, of, searchParams) => {
...
     el.append(notValidUrl);
   };
   el.textContent = '';
   const search = event.target.value;
-  searchParams = new URL(search).searchParams;
+  try {
```
Comment (narcis-radu) [inline libs/blocks/commerce/commerce.js:138]:
```suggestion
  const osi = new URLSearchParams(search).get('osi');
  if (!osi) {
    displaySearchError();
    return;
  }
```

File: libs/blocks/section-metadata/section-metadata.js [hunk 5]
```diff
@@ -53,12 +51,12 @@ function handleStickySection(sticky, section) {
   }
 }
 
-export function handleStyle(text, section) {
+export async function handleStyle(text, section) {
   if (!text || !section) return;
   const styles = text.split(', ').map((style) => style.replaceAll(' ', '-'));
   const sticky = styles.find((style) => style === 'sticky-top' || style === 'sticky-bottom');
   if (sticky) {
```
Comment (narcis-radu) [inline libs/blocks/section-metadata/section-metadata.js:58]:
```suggestion
  if (sticky) await handleStickySection(sticky, section);
```

File: libs/blocks/commerce/commerce.js (hunk 3 above)
Comment (vhargrave) [reply libs/blocks/commerce/commerce.js]: oops, thx

File: libs/blocks/commerce/commerce.js (hunk 4 above)
Comment (vhargrave) [reply libs/blocks/commerce/commerce.js]: love this suggestion! - done

File: libs/blocks/section-metadata/section-metadata.js (hunk 5 above)
Comment (vhargrave) [reply libs/blocks/section-metadata/section-metadata.js]: done!

**Review by narcis-radu — APPROVED** [review]

**Review by vladen — APPROVED** [review]

**Comment by vhargrave** [comment]
> @vhargrave I like the "actions" idea... intersection observer could go here too! Very cool.

@auniverseaway thank you sir - and that's a great idea. I've gone ahead and added a task to the backlog to do that too :) https://jira.corp.adobe.com/browse/MWPW-133762

File: libs/blocks/commerce/commerce.js (hunk 4 above)
Comment (vhargrave) [reply libs/blocks/commerce/commerce.js:137]: @narcis-radu I just discovered that URLSearchParams don't properly if you pass a full URL in as the param. The proposal didn't therefore quite work, but good to know for the future.  

- https://stackoverflow.com/questions/55103757/urlsearchparams-returning-null-for-the-first-query-string
- https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams#no_url_parsing

File: libs/blocks/commerce/commerce.js [hunk 6]
```diff
@@ -127,24 +128,23 @@ export const decorateOfferDetails = async (el, of, searchParams) => {
...
   };
   el.textContent = '';
   const search = event.target.value;
-  searchParams = new URL(search).searchParams;
-  const osi = searchParams.get('osi');
-  if (!osi) {
+  try {
+    const url = new URL(search);
+    const osi = url.searchParams.get('osi');
+    if(!osi) { displaySearchError(); return; }
```
Comment (narcis-radu) [inline libs/blocks/commerce/commerce.js:140]: there's a space missing after `if`

Comment (vhargrave) [reply libs/blocks/commerce/commerce.js:140]: thx!

**Pull request merged**